### ISSUE TITLE
Autogradient of function cannot handle batched forward-prop

### ISSUE THREAD

**Issue by NaimKabir** [twitter/torch-autograd#171]
First of all, autograd is great. Thanks for building it! Makes convoluted nets much easier to write and understand.

Anyway:

To drill down on my issue in it's simplest case:

I'll create two simple neural nets below:

```
lookupfunc, lookuplayers = g.model.NeuralNetwork({
    inputFeatures = 1,
    hiddenFeatures = {64},
    activations = 'Identity',
    classifier = false,
    dropoutProbs = {0}
    })

predictfunc, predictlayers = g.model.NeuralNetwork({
    inputFeatures = 640,
    hiddenFeatures = {1},
    activations = 'Identity',
    classifier = false,
    dropoutProbs = {0}
    })
```
The first net can take in a number and map it to a 64-column Tensor.
The second net can take a 640-column Tensor and map it to a single value answer.
Ok.

In the function that I'll want to apply the autograd to, I want to use the 'lookupfunc' like I do in the following code block. This isn't actually what I'm doing, just the simplest case I can think of :(
```
input = torch.Tensor{1,2,3,4,5,6,7,8,9,10}
weights = {lookuplayers = lookuplayers, predictlayers = predictlayers}
criterion = somefunction()


function forwardpropagation(weights, input, truth)

    vec = lookupfunc(weights.lookuplayers, torch.Tensor{input}) --This is the problem area! Because I'm batching forward propagation and giving these layers 10 separate inputs at once, this thing returns a 10x64 Tensor.

    vec = vec:resize(640) --just getting this into a form that can be taken in by the second net
    
     prediction = predictfunc(weights.predictlayers, vec) --This returns the single number answer

     loss = criterion(prediction, truth)

    return loss
end

```
Forward propagation works fine, and I get a loss.
Now, when I try to do:

```
deltas = autograd(forwardpropagation)

deltas(weights, input, truth)

```

It breaks and returns this stacktrace:

```
        [C]: in function 'addmm'
       ~ /.luarocks/share/lua/5.1/nn/Linear.lua:86: in function 'updateGradInput'
        ~/.luarocks/share/lua/5.1/nn/Module.lua:31: in function 'backward'
        ~/.luarocks/share/lua/5.1/autograd/nnwrapper.lua:222: in function 'nodeApply'
        ~/.luarocks/share/lua/5.1/autograd/nnwrapper.lua:239: in function <~/.luarocks/share/lua/5.1/autograd/nnwrapper.lua:237>
        ...cks/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:186: in function 'gradOnly'
        ...cks/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:229: in function 'deltas'

```

What I noticed after printing out lines inside 'DirectTape' was that it was trying to I think compute gradients for a 10x64 matrix of weights. My hunch is that it got confused by my batch forward propagation when I did this:
`
    vec = lookupfunc(weights.lookuplayers, torch.Tensor{input}) --This returns a 10x64 Tensor
    `

And I confirmed that hunch by instead using a for-loop to substitute that step with this:

```
vec = {}
for element = 1, input:size()[1] do
     table.insert(vec, lookupfunc(weights.lookuplayers, input[element]))
end
vec = torch.cat(vec, 2)
```

This works fine and doesn't confuse AutoGrad at all. However the slowdown seems pretty significant compared to doing a batch forward prop through the layers, so it would be great to be able to fix the autograd issue.

Any support with this strange problem would be appreciated! Thank you.

EDIT:

Digging further into where autograd fails (line 186 of DirectTape):

`  local gradUpdate = (node.gradFun[iarg])(node.outgrad, node.value, table.unpack(node.argValues))`

I've found that the 'node' object that's created for the result of this batch operation has 

- a `node.outgrad` of dimensionality that agrees with my initially defined weights, 
- a `node.value `that does NOT agree with the `node.outgrad` and instead has as many rows as I had batched inputs
- an unpacked `node.argValues` that list the proper weights, but as an input it's listing the full list of batched inputs!

I think the dimensionality disagreements between `node.outgrad` and `node.value` are the cause of the problem, since on function calls of `(node.gradFun[iarg])` where it works, those properties always agree.

**Comment by NaimKabir** [comment]
True issue was unrelated to batching.

Will post a new issue with details as to what went wrong and the hacky solution I had to use to fix it.
